libraries/AP_Mount/AP_Mount_Scripting.cpp

```diff
@@ -0,0 +1,243 @@
+#include "AP_Mount_Scripting.h"
+
+#if HAL_MOUNT_SCRIPTING_ENABLED
+#include <AP_HAL/AP_HAL.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define AP_MOUNT_SCRIPTING_TIMEOUT_MS    1000   // scripting mount becomes unhealthy after 1sec with no updates
+
+#define AP_MOUNT_SCRIPTING_DEBUG 0
+#define debug(fmt, args ...) do { if (AP_MOUNT_SCRIPTING_DEBUG) { GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Siyi: " fmt, ## args); } } while (0)
+
+// update mount position - should be called periodically
+void AP_Mount_Scripting::update()
+{
+    // update based on mount mode
+    switch (get_mode()) {
+        // move mount to a "retracted" position.  To-Do: remove support and replace with a relaxed mode?
+        case MAV_MOUNT_MODE_RETRACT: {
+            const Vector3f &angle_bf_target = _params.retract_angles.get();
+            target_angle_rad.roll = ToRad(angle_bf_target.x);
+            target_angle_rad.pitch = ToRad(angle_bf_target.y);
+            target_angle_rad.yaw = ToRad(angle_bf_target.z);
+            target_angle_rad.yaw_is_ef = false;
+            target_angle_rad_valid = true;
+
+            // mark other targets as invalid
+            target_rate_rads_valid = false;
+            target_loc_valid = false;
+            break;
+        }
+
+        // move mount to a neutral position, typically pointing forward
+        case MAV_MOUNT_MODE_NEUTRAL: {
+            const Vector3f &angle_bf_target = _params.neutral_angles.get();
+            target_angle_rad.roll = ToRad(angle_bf_target.x);
+            target_angle_rad.pitch = ToRad(angle_bf_target.y);
+            target_angle_rad.yaw = ToRad(angle_bf_target.z);
+            target_angle_rad.yaw_is_ef = false;
+            target_angle_rad_valid = true;
+
+            // mark other targets as invalid
+            target_rate_rads_valid = false;
+            target_loc_valid = false;
+            break;
+        }
+
+        // point to the angles given by a mavlink message
+        case MAV_MOUNT_MODE_MAVLINK_TARGETING:
+            switch (mavt_target.target_type) {
+            case MountTargetType::ANGLE:
+                target_angle_rad = mavt_target.angle_rad;
+                target_angle_rad_valid = true;
+                target_rate_rads_valid = false;
+                target_loc_valid = false;
+                break;
+            case MountTargetType::RATE:
+                target_rate_rads = mavt_target.rate_rads;
+                target_rate_rads_valid = true;
+                target_angle_rad_valid = false;
```

Comment:
```suggestion
                target_angle_rad_valid = false;
                target_rate_rads_valid = true;
```

Reply:
Keeping the same order seams more natural.

Reply:
txs for the review.  I think I'll leave it in the current order because the line above sets the target.. then this line says that the target is valid.